File: 1351-count-negative-numbers-in-a-sorted-matrix/1351-count-negative-numbers-in-a-sorted-matrix.js
/**
 * @param {number[][]} grid
 * @return {number}
 */

var countNegatives = function (grid) {
  let answer = 0;
  for (let i = 0; i < grid.length; i++) {
    answer += binarySearch(grid[i]);
  }
  return answer;
};

function binarySearch(arr) {
  let left = 0;
  let right = arr.length - 1;
  let result = 0
  
  if(arr[0] < 0) return arr.length
  if(arr[arr.length-1] >= 0) return 0
    
  while (left <= right) {
    let mid = Math.floor((left + right) / 2);
    if (arr[mid] < 0) {
      result = mid;
      right = mid - 1;
    } else {
      left = mid + 1;
    }
  }

  return arr.length - result
}

File: 1351-count-negative-numbers-in-a-sorted-matrix/1351-count-negative-numbers-in-a-sorted-matrix.java
class Solution {
    public int countNegatives(int[][] grid) {
        int answer = 0;
        
        for(int[] nums: grid){
            answer += nums.length - binarySearch(nums);
        }
        return answer;
    }
    
    public int binarySearch(int[] nums){
        if(nums[0] < 0) return 0;
        if(nums[nums.length-1] >= 0) return nums.length;
        
        int low = 0;
        int high = nums.length - 1;
        int result = 0;
        
        while(low <= high){
            int mid = (int) Math.floor((low + high) / 2);
            
            if(nums[mid] < 0) {
                result = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return result;
    }
}